src/clib/initialize_rates.cpp

```diff
@@ -374,27 +420,30 @@ int grackle::impl::initialize_rates(
 
     //* 1) Rate Coefficients (excluding the external radiation field)
     if (my_chemistry->primordial_chemistry > 0){ 
+        grackle::impl::CollisionalRxnRateCollection* kcol_rate_tables =
+          my_rates->opaque_storage->kcol_rate_tables;
+
         //--------Calculate multispecies collissional rates--------
-        add_reaction_rate(&my_rates->k1, k1_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k3, k3_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k4, k4_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k2, k2_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k5, k5_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k6, k6_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k7, k7_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k8, k8_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k9, k9_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k10, k10_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k11, k11_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k12, k12_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k14, k14_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k15, k15_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k16, k16_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k17, k17_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k18, k18_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k19, k19_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k20, k20_rate, kUnit, my_chemistry);
-        add_reaction_rate(&my_rates->k23, k23_rate, kUnit, my_chemistry);
+        init_preallocated_rate(kcol_rate_tables->data[CollisionalRxnLUT::k1], k1_rate, kUnit, my_chemistry);
+        init_preallocated_rate(kcol_rate_tables->data[CollisionalRxnLUT::k3], k3_rate, kUnit, my_chemistry);
+        init_preallocated_rate(kcol_rate_tables->data[CollisionalRxnLUT::k4], k4_rate, kUnit, my_chemistry);
+        init_preallocated_rate(kcol_rate_tables->data[CollisionalRxnLUT::k2], k2_rate, kUnit, my_chemistry);
```

Comment:
This is not your doing, but maybe now could be the time where we put this in the correct order.